### PR TITLE
Validate signer client options based on authtype

### PR DESCRIPTION
**Pull request by dagnir**
## Motivation and Context
This fixes an issue with the client option validation where the client builder is incorrectly validating that a value for the SIGNER option is present, even when it's not required (because the service does not use AWS auth).

## Modifications
This change only does the appropriate validations within the builder based on the authtypes present in the service model.

## Testing
- New  unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by dave-fn** [comment]
No issues at all, Looks good to me.

**Review by dave-fn — COMMENTED** [review]
No issues at all, Looks good to me.

File: codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/BaseClientBuilderClass.java [hunk 1]
```diff
@@ -623,4 +625,33 @@ private boolean hasClientContextParams() {
...
+                                               .returns(void.class);
+
+        if (AuthUtils.usesAwsAuth(model)) {
+            builder.addStatement("$T.notNull(c.option($T.SIGNER), $S)",
+                                 Validate.class,
+                                 SdkAdvancedClientOption.class,
+                                 "The 'overrideConfiguration.advancedOption[SIGNER]' must be configured in the client builder.");
+        }
+
+        if (AuthUtils.usesBearerAuth(model)) {
```
Comment (zoewangg) [inline codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/BaseClientBuilderClass.java:642]: Should it be `else if`? Can they use both aws auth and bearer auth?

Comment (dagnir) [reply codegen/src/main/java/software/amazon/awssdk/codegen/poet/builder/BaseClientBuilderClass.java:642]: Yes they can. For example, the service `signatureVersion` has `v4` but also includes one operation that uses `bearer`.

**Review by zoewangg — APPROVED** [review]

**Pull request merged**